src/main/java/org/opensearch/security/authtoken/jwt/JwtVendor.java

```diff
@@ -148,4 +152,68 @@ public ExpiringBearerAuthToken createJwt(
 
         return new ExpiringBearerAuthToken(signedJwt.serialize(), subject, expiryTime, expirySeconds);
     }
+
+    @SuppressWarnings("removal")
+    public ExpiringBearerAuthToken createJwt(
+        final String issuer,
+        final String subject,
+        final String audience,
+        final long expiration,
+        final List<String> clusterPermissions,
+        final List<ApiToken.IndexPermission> indexPermissions
+    ) throws JOSEException, ParseException {
+        final long currentTimeMs = timeProvider.getAsLong();
+        final Date now = new Date(currentTimeMs);
+
+        final JWTClaimsSet.Builder claimsBuilder = new JWTClaimsSet.Builder();
+        claimsBuilder.issuer(issuer);
+        claimsBuilder.issueTime(now);
+        claimsBuilder.subject(subject);
+        claimsBuilder.audience(audience);
+        claimsBuilder.notBeforeTime(now);
+
+        final Date expiryTime = new Date(expiration);
+        claimsBuilder.expirationTime(expiryTime);
+
+        if (clusterPermissions != null) {
+            final String listOfClusterPermissions = String.join(",", clusterPermissions);
+            claimsBuilder.claim("cp", encryptString(listOfClusterPermissions));
```

Comment:
Why the use of encryption?

Encryption was used with claims like `backend_roles` and `roles` in obo because we made the choice that roles information was sensitive and should be opaque to extensions. One of the goals was not to propagate some of the existing problems with plugins to the extensions ecosystem. For example, we did not want to propagate roles injection to extensions. If roles injection was supported, then an extension would have been able to send roles (maybe either by http header or transport) to the opensearch cluster to tell it how to authorize. Given that extensions were 3pp and designed to be zero-trust that was not a practice that could be extended to extensions.

Back then we discussed tokens as the strategic way for async tasks for extensions...and we can do the same for plugins as well to eventually deprecate roles injection.

For the extensions use-case, the thinking was that a user can obtain a token (if they have the correct permissions) and then upload it to the extension (via the extension's frontend ui). This was the extension can make calls with the token on the original user's behalf, without the original user needing an active session.

Reply:
Ok I see, in the event a token is leaked the entity obtaining the token should not be able to determine its scope.

I'm not sure if these values need to be included in the payload. We are storing the metadata (including permissions) in a system index and then mirror that system index in a cache for performance. When a client makes a request using a token, we can authenticate it and then look up its permissions in the cache using the jti.